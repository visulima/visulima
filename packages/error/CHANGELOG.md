## @visulima/error [3.2.11](https://github.com/visulima/visulima/compare/@visulima/error@3.2.10...@visulima/error@3.2.11) (2024-06-06)


### Bug Fixes

* allow node v22 ([890d457](https://github.com/visulima/visulima/commit/890d4570f18428e2463944813c0c638b3f142803))



### Dependencies

* **@visulima/path:** upgraded to 1.0.2
* **@visulima/nextra-theme-docs:** upgraded to 4.0.26

## @visulima/error [3.2.10](https://github.com/visulima/visulima/compare/@visulima/error@3.2.9...@visulima/error@3.2.10) (2024-06-05)


### Miscellaneous Chores

* updated dev dependencies ([a2e0504](https://github.com/visulima/visulima/commit/a2e0504dc239049434c2482756ff15bdbaac9b54))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.25

## @visulima/error [3.2.9](https://github.com/visulima/visulima/compare/@visulima/error@3.2.8...@visulima/error@3.2.9) (2024-05-24)


### Bug Fixes

* changed pathe to @visulima/path ([#410](https://github.com/visulima/visulima/issues/410)) ([bfe1287](https://github.com/visulima/visulima/commit/bfe1287aff6d28d5dca302fd4d58c1f6234ce0bb))


### Miscellaneous Chores

* changed semantic-release-npm to pnpm ([b6d100a](https://github.com/visulima/visulima/commit/b6d100a2bf3fd026577be48726a37754947f0973))



### Dependencies

* **@visulima/path:** upgraded to 1.0.1

## @visulima/error [3.2.8](https://github.com/visulima/visulima/compare/@visulima/error@3.2.7...@visulima/error@3.2.8) (2024-05-15)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.24

## @visulima/error [3.2.7](https://github.com/visulima/visulima/compare/@visulima/error@3.2.6...@visulima/error@3.2.7) (2024-04-27)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.23

## @visulima/error [3.2.6](https://github.com/visulima/visulima/compare/@visulima/error@3.2.5...@visulima/error@3.2.6) (2024-04-17)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.22

## @visulima/error [3.2.5](https://github.com/visulima/visulima/compare/@visulima/error@3.2.4...@visulima/error@3.2.5) (2024-04-09)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.21

## @visulima/error [3.2.4](https://github.com/visulima/visulima/compare/@visulima/error@3.2.3...@visulima/error@3.2.4) (2024-03-30)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.20

## @visulima/error [3.2.3](https://github.com/visulima/visulima/compare/@visulima/error@3.2.2...@visulima/error@3.2.3) (2024-03-27)


### Bug Fixes

* added missing os key to package.json ([4ad1268](https://github.com/visulima/visulima/commit/4ad1268ed12cbdcf60aeb46d4c052ed1696bc150))

## @visulima/error [3.2.2](https://github.com/visulima/visulima/compare/@visulima/error@3.2.1...@visulima/error@3.2.2) (2024-03-22)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.19

## @visulima/error [3.2.1](https://github.com/visulima/visulima/compare/@visulima/error@3.2.0...@visulima/error@3.2.1) (2024-03-16)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.18

## @visulima/error [3.2.0](https://github.com/visulima/visulima/compare/@visulima/error@3.1.2...@visulima/error@3.2.0) (2024-03-11)


### Features

* added new index-to-line-column function ([c794be0](https://github.com/visulima/visulima/commit/c794be0867eae45ce66096bd15691d7437545a5d))

## @visulima/error [3.1.2](https://github.com/visulima/visulima/compare/@visulima/error@3.1.1...@visulima/error@3.1.2) (2024-03-10)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.17

## @visulima/error [3.1.1](https://github.com/visulima/visulima/compare/@visulima/error@3.1.0...@visulima/error@3.1.1) (2024-03-09)


### Bug Fixes

* **error:** fixed color option override ([467f1dc](https://github.com/visulima/visulima/commit/467f1dc979b6d4bca102589d330be083fd44e43e))

## @visulima/error [3.1.0](https://github.com/visulima/visulima/compare/@visulima/error@3.0.7...@visulima/error@3.1.0) (2024-03-09)


### Features

* **error:** added new tabWidth option to disable the tab transformer ([a115918](https://github.com/visulima/visulima/commit/a11591825f841c9ab28ebadd85d20633c5cd9e50))

## @visulima/error [3.0.7](https://github.com/visulima/visulima/compare/@visulima/error@3.0.6...@visulima/error@3.0.7) (2024-03-09)


### Bug Fixes

* **error:** added missing CodeFrameLocation export ([d4144e0](https://github.com/visulima/visulima/commit/d4144e0bd579797c49c9fe5b867289ef89400293))

## @visulima/error [3.0.6](https://github.com/visulima/visulima/compare/@visulima/error@3.0.5...@visulima/error@3.0.6) (2024-03-06)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.16

## @visulima/error [3.0.5](https://github.com/visulima/visulima/compare/@visulima/error@3.0.4...@visulima/error@3.0.5) (2024-03-04)


### Bug Fixes

* fixed all found type issues ([eaa40d1](https://github.com/visulima/visulima/commit/eaa40d11f3fc056dfddcc25404bf109587ef2862))
* minifyWhitespace on prod build, removed @tsconfig/* configs ([410cb73](https://github.com/visulima/visulima/commit/410cb737c44c445a0479bdd49b4100d5daf2d83d))
* updated @jridgewell/trace-mapping ([2955e4a](https://github.com/visulima/visulima/commit/2955e4adad7478947e4b224fad7bca9c8d5fe4d5))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.15

## @visulima/error [3.0.4](https://github.com/visulima/visulima/compare/@visulima/error@3.0.3...@visulima/error@3.0.4) (2024-02-26)


### Bug Fixes

* updated @jridgewell/trace-mapping ([f2061d1](https://github.com/visulima/visulima/commit/f2061d11c0db0df18b1ae065118e6a2f33811eab))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.14

## @visulima/error [3.0.3](https://github.com/visulima/visulima/compare/@visulima/error@3.0.2...@visulima/error@3.0.3) (2024-01-31)


### Bug Fixes

* update deps ([1b01d6f](https://github.com/visulima/visulima/commit/1b01d6f55435ac2ed67db48e6e2060f786f59e56))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.13

## @visulima/error [3.0.2](https://github.com/visulima/visulima/compare/@visulima/error@3.0.1...@visulima/error@3.0.2) (2024-01-19)


### Bug Fixes

* updated all deps, updated test based on eslint errors ([909f8f3](https://github.com/visulima/visulima/commit/909f8f384804d7ef140354ab44f867532dbc9847))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.12

## @visulima/error [3.0.1](https://github.com/visulima/visulima/compare/@visulima/error@3.0.0...@visulima/error@3.0.1) (2023-12-06)


### Bug Fixes

* fixed possible deep nesting of error cause in getErrorCauses ([350ddba](https://github.com/visulima/visulima/commit/350ddbaf2b8120a0815f1673fc1582ca7946847d))

## @visulima/error [3.0.0](https://github.com/visulima/visulima/compare/@visulima/error@2.0.0...@visulima/error@3.0.0) (2023-12-06)


### ⚠ BREAKING CHANGES

* removed ErrorWithMetadata, fixed wrong Trace on error namespace
Signed-off-by: prisis <d.bannert@anolilab.de>

### Features

* added new getErrorCauses helper ([e94d9ee](https://github.com/visulima/visulima/commit/e94d9ee34d6fa0f93d853a227e9b31bab4aacdab))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.11

## @visulima/error [2.0.0](https://github.com/visulima/visulima/compare/@visulima/error@1.1.1...@visulima/error@2.0.0) (2023-12-04)


### ⚠ BREAKING CHANGES

* removed positionAt, changed public interface of code-frame
Signed-off-by: prisis <d.bannert@anolilab.de>

### Features

* changed public interface of code-frame, added multiline marker ([d9d60c0](https://github.com/visulima/visulima/commit/d9d60c06029f7564bc956a027d040331eddd0443))


### Bug Fixes

* fixed wrong export ([b3689cd](https://github.com/visulima/visulima/commit/b3689cd5f5479cf934d8d892ff7071feedd3b9ae))

## @visulima/error [1.1.1](https://github.com/visulima/visulima/compare/@visulima/error@1.1.0...@visulima/error@1.1.1) (2023-11-30)



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.10

## @visulima/error [1.1.0](https://github.com/visulima/visulima/compare/@visulima/error@1.0.2...@visulima/error@1.1.0) (2023-11-30)


### Features

* added a stacktrace parser and a sourcemap loader ([#250](https://github.com/visulima/visulima/issues/250)) ([4543e44](https://github.com/visulima/visulima/commit/4543e44becc85a54cdbe7701e2c3e6601b282985))

## @visulima/error [1.0.2](https://github.com/visulima/visulima/compare/@visulima/error@1.0.1...@visulima/error@1.0.2) (2023-11-30)


### Bug Fixes

* **deps:** updated package deps ([b4f4ede](https://github.com/visulima/visulima/commit/b4f4eded7bbded62b341ade0017ab357336f3af2))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.9

## @visulima/error [1.0.1](https://github.com/visulima/visulima/compare/@visulima/error@1.0.0...@visulima/error@1.0.1) (2023-11-07)


### Bug Fixes

* fixed the homepage url of the package ([02075ce](https://github.com/visulima/visulima/commit/02075ce997d62c1caf79690b32dd2f931e64bebe))



### Dependencies

* **@visulima/nextra-theme-docs:** upgraded to 4.0.8

## @visulima/error 1.0.0 (2023-11-07)


### Features

* new error package ([#248](https://github.com/visulima/visulima/issues/248)) ([89b76f5](https://github.com/visulima/visulima/commit/89b76f5a455363bfc3852759d4e5f64d71b48925))
