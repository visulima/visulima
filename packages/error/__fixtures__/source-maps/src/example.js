import React, {PropTypes, PureComponent} from 'react'

export default class SomeComponent extends PureComponent{
  static propTypes = {
    className: PropTypes.string
  }

  constructor(props) {
    super(props)

    this.state = {count: 0}
  }

  handleClick = () => {
    this.setState(prevState => ({count: prevState.count + 1}))
  }

  render() {
    return (
      <button
        className={this.props.className}
        onClick={this.handleClick}>
        Clicked {this.state.count} times
      </button>
    )
  }
}
