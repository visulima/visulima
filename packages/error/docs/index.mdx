---
title: Introduction
description: A comprehensive error handling library for Node.js and modern JavaScript applications
---

import { Cards, Card } from 'fumadocs-ui/components/card';
import { Callout } from 'fumadocs-ui/components/callout';

## Overview

`@visulima/error` provides advanced error handling capabilities beyond standard JavaScript errors, including:

- Custom error classes with enhanced properties
- Stack trace parsing and formatting
- Code frame generation for error locations
- Error serialization and deserialization
- AI-powered error solutions
- Pretty error rendering for better debugging

<Callout type="info">
This package supports Node.js 18+ and all modern browsers (excluding those older than 6 years).
</Callout>

## Quick Start

Get started with the basics:

<Cards>
  <Card 
    href="/docs/error/installation" 
    title="Installation" 
    description="How to install and set up the package"
  />
  <Card 
    href="/docs/error/quick-start" 
    title="Quick Start" 
    description="Get up and running with basic examples"
  />
  <Card 
    href="/docs/error/examples" 
    title="Examples" 
    description="Practical examples and use cases"
  />
</Cards>

## Core Features

Explore the main functionality:

<Cards>
  <Card 
    href="/docs/error/visulima-error" 
    title="VisulimaError Class" 
    description="Custom error class with enhanced features"
  />
  <Card 
    href="/docs/error/stacktrace" 
    title="Stack Trace Parsing" 
    description="Parse and format error stack traces"
  />
  <Card 
    href="/docs/error/code-frames" 
    title="Code Frames" 
    description="Display code snippets with error locations"
  />
  <Card 
    href="/docs/error/serialization" 
    title="Error Serialization" 
    description="Serialize and deserialize errors"
  />
  <Card 
    href="/docs/error/rendering" 
    title="Error Rendering" 
    description="Pretty print errors with colors"
  />
  <Card 
    href="/docs/error/solution-finders" 
    title="Solution Finders" 
    description="Generate helpful hints to fix errors"
  />
</Cards>

## Advanced Topics

<Cards>
  <Card 
    href="/docs/error/ai-integration" 
    title="AI Integration" 
    description="Use AI to generate error solutions"
  />
  <Card 
    href="/docs/error/custom-errors" 
    title="Custom Error Types" 
    description="Create your own error classes"
  />
  <Card 
    href="/docs/error/typescript" 
    title="TypeScript Support" 
    description="TypeScript types and interfaces"
  />
</Cards>

## Requirements

- Node.js >= 18.0.0
- TypeScript >= 5.0 (for TypeScript projects)

## Links

- [GitHub Repository](https://github.com/visulima/visulima/tree/main/packages/error)
- [NPM Package](https://www.npmjs.com/package/@visulima/error)
- [Report Issues](https://github.com/visulima/visulima/issues)
