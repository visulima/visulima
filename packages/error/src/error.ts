export type { CodeFrameOptions, ErrorLocation, ErrorProperties, ErrorWithMetadata, Trace } from "./types";
export { isVisulimaError, VisulimaError } from "./visulima-error";
