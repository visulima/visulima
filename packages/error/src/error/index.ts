export { default as getErrorCauses } from "./get-error-causes";
export type { Options as RenderErrorOptions } from "./render/error";
export { renderError } from "./render/error";
export type { SerializedError } from "./serialize/error-proto";
export type { Options as ErrorWithCauseSerializerOptions } from "./serialize/serialize";
export { serialize as serializeError } from "./serialize/serialize";
export type { ErrorHint, ErrorLocation, ErrorProperties } from "./types";
export { isVisulimaError, VisulimaError } from "./visulima-error";
