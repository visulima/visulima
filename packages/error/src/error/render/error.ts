import { existsSync, readFileSync } from "node:fs";
import { relative } from "node:path";
import { cwd } from "node:process";
import { fileURLToPath } from "node:url";

import type { CodeFrameOptions, ColorizeMethod } from "../../code-frame";
import { codeFrame } from "../../code-frame";
import type { Trace } from "../../stacktrace";
import { parseStacktrace } from "../../stacktrace";
import type { VisulimaError } from "../visulima-error";

const getPrefix = (indentation: number | "\t", deep: number): string => {
    if (deep === 0) {
        return "";
    }

    if (indentation === "\t") {
        return "\t".repeat(deep);
    }

    return " ".repeat(indentation * deep);
};

const getRelativePath = (filePath: string, cwdPath: string) => {
    /**
     * Node.js error stack is all messed up. Some lines have file info
     * enclosed in parentheses and some are not
     */
    const path = filePath.replace("async file:", "file:");

    return relative(cwdPath, path.startsWith("file:") ? fileURLToPath(path) : path);
};

/**
 * Returns the error message
 */
const getMessage = (error: AggregateError | Error | VisulimaError, { color, hideErrorTitle, indentation }: Options, deep: number): string =>
    getPrefix(indentation, deep) + (hideErrorTitle ? color.title(error.message) : color.title(error.name + ": " + error.message)) + "\n";

const getHint = (error: AggregateError | Error | VisulimaError, { color, indentation }: Options, deep: number): string | undefined => {
    if ((error as VisulimaError).hint === undefined) {
        return undefined;
    }

    const prefix = getPrefix(indentation, deep);

    let message = "";

    if (Array.isArray((error as VisulimaError).hint)) {
        // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax
        for (const line of (error as VisulimaError).hint as string[]) {
            message += prefix + line + "\n";
        }
    } else {
        message += prefix + ((error as VisulimaError).hint as string);
    }

    return color.hint(message);
};

const getMainFrame = (trace: Trace, { color, cwd: cwdPath, displayShortPath, indentation }: Options, deep = 0): string => {
    const filePath = displayShortPath ? getRelativePath(trace.file as string, cwdPath) : trace.file;

    const { fileLine, method } = color;

    return (
        getPrefix(indentation, deep) +
        "at " +
        (trace.methodName ? method(trace.methodName) + " " : "") +
        fileLine(filePath as string) +
        ":" +
        fileLine(trace.line + "")
    );
};

const getCode = (
    trace: Trace,
    { color, indentation, linesAbove, linesBelow, showGutter, showLineNumbers, tabWidth }: Options,
    deep: number,
): string | undefined => {
    if (trace.file === undefined) {
        return undefined;
    }

    const filePath = trace.file.replace("file://", "");

    // eslint-disable-next-line security/detect-non-literal-fs-filename
    if (!existsSync(filePath)) {
        return undefined;
    }

    // eslint-disable-next-line security/detect-non-literal-fs-filename
    const fileContent = readFileSync(filePath, "utf8");

    return codeFrame(
        fileContent,
        {
            start: { column: trace.column, line: trace.line as number },
        },
        { color, linesAbove, linesBelow, prefix: getPrefix(indentation, deep), showGutter, showLineNumbers, tabWidth },
    );
};

const getErrors = (error: AggregateError, options: Options, deep: number): string | undefined => {
    if (error.errors.length === 0) {
        return undefined;
    }

    const prefix = getPrefix(options.indentation, deep);

    let message = prefix + "Errors:\n\n";
    let first = true;

    // eslint-disable-next-line no-loops/no-loops,no-restricted-syntax,@typescript-eslint/naming-convention,no-underscore-dangle
    for (const error_ of error.errors) {
        if (first) {
            first = false;
        } else {
            message += "\n\n";
        }

        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        message += internalRenderError(
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
            error_,
            { ...options, framesMaxLimit: 0, hideErrorCodeView: options.hideErrorErrorsCodeView },
            deep + 1,
        );
    }

    return "\n" + message;
};

const getCause = (error: AggregateError | Error | VisulimaError, options: Options, deep: number): string => {
    let message = getPrefix(options.indentation, deep) + "Caused by:\n\n";

    const cause = error.cause as Error;

    message += getMessage(cause, options, deep);

    const stacktrace = parseStacktrace(cause);
    const mainFrame = stacktrace.shift() as Trace;

    const hint = getHint(cause, options, deep);

    if (hint) {
        message += hint + "\n";
    }

    message += getMainFrame(mainFrame, options, deep);

    if (!options.hideErrorCauseCodeView) {
        message += "\n" + getCode(mainFrame, options, deep);
    }

    if (cause.cause) {
        message += "\n" + getCause(cause, options, deep + 1);
    } else if (cause instanceof AggregateError) {
        message += "\n" + getErrors(cause, options, deep);
    }

    return "\n" + message;
};

const getStacktrace = (stack: Trace[], options: Options): string => {
    const frames = stack.slice(0, options.framesMaxLimit);

    return (frames.length > 0 ? "\n" : "") + frames.map((frame) => getMainFrame(frame, options)).join("\n");
};

const internalRenderError = (error: AggregateError | Error | VisulimaError, options: Partial<Options>, deep: number): string => {
    const config = {
        cwd: cwd(),
        displayShortPath: false,
        framesMaxLimit: Number.POSITIVE_INFINITY,
        hideErrorCauseCodeView: false,
        hideErrorCodeView: false,
        hideErrorErrorsCodeView: false,
        hideErrorTitle: false,
        hideMessage: false,
        indentation: 4,
        linesAbove: 2,
        linesBelow: 3,
        showGutter: true,
        showLineNumbers: true,
        tabWidth: 4,
        ...options,
        color: {
            fileLine: (value: string) => value,
            gutter: (value: string) => value,
            hint: (value: string) => value,
            marker: (value: string) => value,
            message: (value: string) => value,
            method: (value: string) => value,
            title: (value: string) => value,
            ...options.color,
        },
    } satisfies Options;

    const stack = parseStacktrace(error);
    const mainFrame = stack.shift();

    return [
        options.hideMessage ? undefined : getMessage(error, config, deep),
        "",
        getHint(error, config, deep),
        "",
        mainFrame ? getMainFrame(mainFrame, config, deep) : undefined,
        mainFrame && !config.hideErrorCodeView ? getCode(mainFrame, config, deep) : undefined,
        error instanceof AggregateError ? getErrors(error, config, deep) : undefined,
        error.cause === undefined ? undefined : getCause(error, config, deep),
        stack.length > 0 ? getStacktrace(stack, config) : undefined,
    ]
        .filter(Boolean)
        .join("\n");
};

export type Options = Omit<CodeFrameOptions, "message | prefix"> & {
    color: CodeFrameOptions["color"] & {
        fileLine: ColorizeMethod;
        hint: ColorizeMethod;
        method: ColorizeMethod;
        title: ColorizeMethod;
    };
    cwd: string;
    displayShortPath: boolean;
    framesMaxLimit: number;
    hideErrorCauseCodeView: boolean;
    hideErrorCodeView: boolean;
    hideErrorErrorsCodeView: boolean;
    hideErrorTitle: boolean;
    hideMessage: boolean;
    indentation: number | "\t";
};

export const renderError = (error: AggregateError | Error | VisulimaError, options: Partial<Options> = {}): string => internalRenderError(error, options, 0);
