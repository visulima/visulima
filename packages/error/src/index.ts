export { default as codeFrame } from "./code-frame";
export type { CodeFrameOptions, ErrorLocation, ErrorProperties, ErrorWithMetadata, Trace } from "./error";
export { isVisulimaError, VisulimaError } from "./error";
export type { TraceMap } from "./sourcemap";
export { generatedPositionFor, loadSourceMap, originalPositionFor, sourceContentFor, traceSegment } from "./sourcemap";
export { parseStacktrace } from "./stacktrace";
export { positionAt } from "./utils";
