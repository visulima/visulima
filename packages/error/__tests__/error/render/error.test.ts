import { describe, expect, it, vi } from "vitest";

import { VisulimaError } from "../../../src";
import { renderError } from "../../../src/error/render/error";

const hoisted = vi.hoisted(() => {
    return {
        existsSync: vi.fn().mockReturnValue(true),
    };
});

vi.mock("node:fs", async () => {
    const original = await vi.importActual("node:fs");

    return {
        ...original,
        existsSync: hoisted.existsSync,
    };
});

describe("renderError", () => {
    it("should render error message with default options", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");
        const errorOutput = renderError(error);
        expect(errorOutput).toContain("Error: This is an example error message");
    });

    it("should render error message with custom color options", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");
        const options = {
            color: {
                fileLine: (value: string) => `\u001B[35m${value}\u001B[39m`,
                hint: (value: string) => `\u001B[33m${value}\u001B[39m`,
                method: (value: string) => `\u001B[36m${value}\u001B[39m`,
                title: (value: string) => `\u001B[31m${value}\u001B[39m`,
            },
        };
        const errorOutput = renderError(error, options);
        expect(errorOutput).toContain("\u001B[31mError: This is an example error message\u001B[39m");
    });

    it("should render error message with short file paths", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");
        const options = { displayShortPath: true };
        const errorOutput = renderError(error, options);

        expect(errorOutput).toContain("at ");
    });

    it("should render error message with limited stack frames", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");

        const errorOutput = renderError(error, { framesMaxLimit: 1, hideErrorCodeView: true });
        const stackFrames = (errorOutput.match(/at <unknown> /g) as RegExpMatchArray).length;

        expect(stackFrames).toBeLessThanOrEqual(2);
    });

    it("should render error message with hidden error title", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");
        const options = { hideErrorTitle: true };
        const errorOutput = renderError(error, options);

        expect(errorOutput).not.toContain("Error:");
    });

    it("should render error message when stack trace is empty", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");

        vi.spyOn(error, "stack", "get").mockReturnValue("");

        const errorOutput = renderError(error);

        expect(errorOutput).toContain("Error: This is an example error message");
    });

    it("should render error message when file does not exist", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");

        hoisted.existsSync.mockReturnValue(false);

        const errorOutput = renderError(error);

        expect(errorOutput).toContain("Error: This is an example error message");
    });

    it("should render error message when error has no hint", () => {
        expect.assertions(1);

        const visulimaError = new VisulimaError({ hint: undefined, message: "This is an example error message", name: "Error" });
        const errorOutput = renderError(visulimaError, {
            hideErrorCodeView: true,
        });

        expect(errorOutput).not.toContain("Hint:");
    });

    it("should render error message when error has no cause", () => {
        expect.assertions(1);

        const visulimaError = new VisulimaError({ cause: undefined, message: "This is an example error message", name: "Error" });
        const errorOutput = renderError(visulimaError, {
            hideErrorCodeView: true,
        });

        expect(errorOutput).not.toContain("Caused by:");
    });

    it("should render error without code view", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message");
        const errorOutput = renderError(error, {
            hideErrorCodeView: true,
        });

        expect(errorOutput).not.toContain("const errorOutput = renderError(error,");
    });

    it("should render error without cause code view", () => {
        expect.assertions(1);

        const error = new Error("This is an example error message", {
            cause: new Error("This is an example cause error message"),
        });
        const errorOutput = renderError(error, {
            hideErrorCauseCodeView: true,
        });

        const found = (errorOutput.match(/This is an example cause error message/g) as RegExpMatchArray).length;

        // Its twice because of the code view
        expect(found).toBeLessThanOrEqual(2);
    });

    it.skipIf(!global.AggregateError)("should render error messages for multiple errors in AggregateError", () => {
        expect.assertions(3);

        const error1 = new Error("Error 1");
        const error2 = new Error("Error 2");
        // eslint-disable-next-line unicorn/error-message
        const aggregateError = new AggregateError([error1, error2]);

        const errorsOutput = renderError(aggregateError);

        expect(errorsOutput).toContain("Errors:");
        expect(errorsOutput).toContain("Error 1");
        expect(errorsOutput).toContain("Error 2");
    });

    it.skipIf(!global.AggregateError)("should handle nested errors within AggregateError", () => {
        expect.assertions(1);

        // eslint-disable-next-line unicorn/error-message
        const nestedError = new AggregateError([new Error("Nested Error")]);
        // eslint-disable-next-line unicorn/error-message
        const errorsOutput = renderError(new AggregateError([nestedError]), {
            displayShortPath: true,
            framesMaxLimit: 1,
            hideErrorCauseCodeView: true,
            hideErrorCodeView: true,
            hideErrorErrorsCodeView: true,
        });
        expect(errorsOutput).toMatchSnapshot();
    });

    it("should handle nested cause errors", () => {
        expect.assertions(1);

        const errorsOutput = renderError(
            new Error("This is an error message", {
                cause: new Error("This is the cause of the error", {
                    cause: new Error("This is the cause of the cause of the error"),
                }),
            }),
            {
                displayShortPath: true,
                framesMaxLimit: 1,
                hideErrorCauseCodeView: true,
                hideErrorCodeView: true,
                hideErrorErrorsCodeView: true,
            },
        );

        expect(errorsOutput).toMatchSnapshot();
    });

    it("should respect base prefix option", () => {
        expect.assertions(1);

        const errorsOutput = renderError(
            new Error("This is an error message", {
                cause: new Error("This is the cause of the error", {
                    cause: new Error("This is the cause of the cause of the error"),
                }),
            }),
            {
                displayShortPath: true,
                framesMaxLimit: 1,
                hideErrorCauseCodeView: true,
                hideErrorCodeView: true,
                hideErrorErrorsCodeView: true,
                prefix: "prefix",
            },
        );

        expect(errorsOutput).toMatchSnapshot();
    });
});
