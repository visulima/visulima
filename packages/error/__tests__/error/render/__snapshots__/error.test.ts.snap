// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderError > should handle nested cause errors 1`] = `
"Error: This is an error message

at <unknown> __tests__/error/render/error.test.ts:187

Caused by:

Error: This is the cause of the error
at <unknown> __tests__/error/render/error.test.ts:188

    Caused by:

    Error: This is the cause of the cause of the error
    at <unknown> __tests__/error/render/error.test.ts:189"
`;

exports[`renderError > should handle nested errors within AggregateError 1`] = `
"AggregateError: 

at <unknown> __tests__/error/render/error.test.ts:171

Errors:

    AggregateError: 

    at <unknown> __tests__/error/render/error.test.ts:169

    Errors:

        Error: Nested Error

        at <unknown> __tests__/error/render/error.test.ts:169"
`;
